src/odemis/gui/comp/overlay/world.py

```diff
@@ -832,6 +833,604 @@ def deactivate(self):
+
+        self.last_shiftscale = None  # previous shift & scale of the canvas to know whether it has changed
+
+        self.mode = NONE_RULER_MODE
+        self._edges = {}  # the bound-boxes of the rulers in view coordinates
+        self.pixel_ruler = False  # true if the ruler is 1 pixel or less
+
+    def __str__(self):
+        return "Ruler %g,%g -> %g,%g" % (self.p_start_pos[0], self.p_start_pos[1],
+                                         self.p_end_pos[0], self.p_end_pos[1])
+
+    def _view_to_phys(self):
+        """ Update the physical position to reflect the view position """
+
+        if self.v_start_pos and self.v_end_pos:
+            offset = self.cnvs.get_half_buffer_size()
+            self.p_start_pos = self.cnvs.view_to_phys(self.v_start_pos, offset)
+            self.p_end_pos = self.cnvs.view_to_phys(self.v_end_pos, offset)
+            self._calc_edges()
+
+    def start_dragging(self, drag, vpos):
+        """
+        The user can start dragging (creating/editing/moving) the ruler when the left mouse button is pressed down
+        Args:
+            drag: hover mode (HOVER_START, HOVER_LINE, HOVER_END)
+            vpos: the view coordinates of the mouse cursor once left click mouse event is fired
+
+        """
+
+        self.drag_v_start_pos = self.drag_v_end_pos = vpos
+
+        if drag == gui.HOVER_START:
+            self.edit_hover = drag
+            self.mode = EDIT_START_RULER_MODE
+        elif drag == gui.HOVER_END:
+            self.edit_hover = drag
+            self.mode = EDIT_END_RULER_MODE
+        elif drag == gui.HOVER_LINE:
+            self.mode = MOVE_RULER_MODE
+        else:
+            raise ValueError("No valid hover mode")
+        self._view_to_phys()
+
+    def on_motion(self, vpos, ctrl_down):
+        """
+        Given that the left mouse button is already pressed down and the mouse cursor is over the ruler,
+        the user can drag (create/edit/move) the ruler until the left button is released.
+        Args:
+            vpos: the view coordinates of the mouse cursor while dragging
+            ctrl_down (boolean): if True, the ctrl key is pressed while dragging and the ruler
+            is forced to be at one angle multiple of 45 degrees.
+        """
+
+        self.drag_v_end_pos = vpos
+
+        if self.mode != NONE_RULER_MODE:
+            if self.mode in (EDIT_START_RULER_MODE, EDIT_END_RULER_MODE):
+                self._update_editing(ctrl_down)
+            else:  # self.mode == MOVE_RULER_MODE:
+                self._update_moving()
+            self.cnvs.Refresh()
+            self._view_to_phys()
+
+    def _update_editing(self, round_angle):
+        """Update view coordinates while editing the ruler. If round_angle(boolean) is True,
+        the ruler is forced to be at one angle multiple of 45 degrees"""
+        current_pos = Vec(self.cnvs.clip_to_viewport(self.drag_v_end_pos))
+
+        if self.edit_hover == gui.HOVER_START:
```

Comment:
Is it possible to just check whether the mode is EDIT_START_RULER_MODE vs EDIT_END_RULER_MODE, and drop the use of `edit_hover`?